### PR TITLE
Fix contrast and accumulation averaging

### PR DESCRIPTION
**Pull request by Hpn4**
The new behavior is:
- On pipe refresh the accumulation queue is cleared.
- The autocontrast is applied (if enabled) if there is no accumulation, at the first image of the accumulation queue (to calibrate the contrast) and when the queue is full (to fine tune the contrast).
- The accumulation averaging is not done on all images but only on the images in the queue (no more black screen).

### CODE REVIEW

File: Holovibes/sources/compute/rendering.cc [hunk 1]
```diff
@@ -227,10 +236,28 @@ void Rendering::insert_apply_contrast(WindowKind view)
...
-                                            std::atomic<bool>& autocontrast_slice_xz_request,
-                                            std::atomic<bool>& autocontrast_slice_yz_request,
-                                            std::atomic<bool>& autocontrast_filter2d_request)
+/*! \brief Tell if the contrast should be applied
+ *
+ * \param request The request for autocontrast
+ * \param queue The accumulation queue
+ * \return true if the contrast should be applied
+ */
+inline bool apply_contrast(bool request, const std::unique_ptr<Queue>& queue)
```
Comment (Baschtien) [inline Holovibes/sources/compute/rendering.cc:245]: Maybe the name is a bit misleading; for me the function should have a name that reflects its 'testing' purpose
Like 'check_autocontrast' or 'do_apply_autocontrast'

**Review by Baschtien — APPROVED** [review]

File: Holovibes/cuda_sources/tools_conversion.cu [hunk 2]
```diff
@@ -417,26 +417,24 @@ void uchar_to_shifted_uchar(uchar* output, const uchar* input, const size_t size
...
 {
     const uint index = blockIdx.x * blockDim.x + threadIdx.x;
-    long int pos = end; // end is excluded
+    long int pos = start;
 
     if (index < nb_pixel)
     {
         float val = 0;
-        for (size_t i = 0; i < nb_elmt; i++)
+        for (size_t i = 0; i < nb_elmt; ++i)
```
[the remark this answers is not in the document]
Comment (ArthurCourselle) [reply Holovibes/cuda_sources/tools_conversion.cu:431]: Team i++

File: Holovibes/sources/compute/rendering.cc (hunk 1 above)
Comment (ArthurCourselle) [inline Holovibes/sources/compute/rendering.cc:245]: Do not forget about [in] for doxgygen, also the inlined functions should be in the header file. I agree with Bastien that the name is unclear.

File: Holovibes/sources/compute/rendering.cc [hunk 3]
```diff
@@ -242,42 +269,52 @@ void Rendering::insert_compute_autocontrast(std::atomic<bool>& autocontrast_requ
         if (!time_transformation_env_.gpu_time_transformation_queue->is_full())
             return;
 
-        if (autocontrast_request &&
-            (!image_acc_env_.gpu_accumulation_xy_queue || image_acc_env_.gpu_accumulation_xy_queue->is_full()))
+        if (apply_contrast(autocontrast_xy_, image_acc_env_.gpu_accumulation_xy_queue))
         {
             // FIXME Handle composite size, adapt width and height (frames_res =
             // buffers_.gpu_postprocess_frame_size)
```
Comment (ArthurCourselle) [inline Holovibes/sources/compute/rendering.cc:275]: Maybe you could remove this fixme , it looks like an old comment

**Review by ArthurCourselle — DISMISSED** [review]
Looks great for me and way more logical that what was done before.
The only changes I request is on the function `apply_contrast`, changing the name, put it in header and add params[in] in doxygen documentation.

**Review by GustaveHerve — APPROVED** [review]

File: Holovibes/sources/compute/rendering.cc (hunk 1 above)
Comment (Baschtien) [reply Holovibes/sources/compute/rendering.cc:245]: Should the function even be inline ? it feels a bit weird having a more than 1 line inline function...

Comment (Baschtien) [reply Holovibes/sources/compute/rendering.cc:245]: Renamed the function to `should_apply_contrast` (inspired by the doxygen description, which contains the word 'should')

File: Holovibes/sources/compute/rendering.cc (hunk 3 above)
Comment (Baschtien) [reply Holovibes/sources/compute/rendering.cc]: Removed

**Review by ArthurCourselle — APPROVED** [review]

**Pull request merged**